### PR TITLE
Extract closure functions to a private method, so that `cocotb.coroutine.__new__` is not invoked each time

### PR DESCRIPTION
**Pull request by eric-wieser**
Not at all clear if this has any real impact, although I'd be surprised if this made things worse.

gh-961 again, pinging @andresdemski since they have a possibly representative testsuite to benchmark.

@imphil, might it be worth creating a "performance" label?

### CODE REVIEW

File: cocotb/triggers.py [hunk 1]
```diff
@@ -554,6 +556,18 @@ class First(_AggregateWaitable):
             t2 = Timer(10, units='ps')
             t_ret = yield First(t1, t2)
     """
+    @staticmethod
+    @decorators.coroutine
+    def __internal_waiter(t, e, completed):
```
Comment (imphil) [inline cocotb/triggers.py]: In the previous code, `t` and `e` were pretty obvious, but with it being a separate function could we go for more descriptive names (event, trigger, completed_triggers)?

Comment (eric-wieser) [reply cocotb/triggers.py]: Seems reasonable, will wait for performance numbers before changing

**Comment by andresdemski** [comment]
* eric-wieser:avoid-runtime-coroutine-creation: SIM / REAL TIME : 21994.61 NS/S
* cocotb:master: SIM / REAL TIME : 21447.10 NS/S

Improvement : ~2,5%

**Comment by imphil** [comment]
Thanks for your data @andresdemski!

**Comment by andresdemski** [comment]
Thank you @imphil and @eric-wieser for your amazing work. I'm just trying to contribute with a little grain of sand.

**Comment by imphil** [comment]
So performance looks good, let me know @eric-wieser when you're done with renaming the arguments and I can merge.

**Comment by eric-wieser** [comment]
@imphil: ended up deciding the closures were worth it for readability - I've rewritten this patch to still avoid the `@coroutine` overhead, without losing the locality of the local function..

@andresdemski, would you mind profiling again? Thanks again for the feedback you've given so far.

**Comment by andresdemski** [comment]
Yes, sure.
[profiling.zip](https://github.com/potentialventures/cocotb/files/3350130/profiling.zip)

# Current master
current_master

# eric-wieser:avoid-runtime-coroutine-creation
avoid-runtime-cro-creation

**Comment by andresdemski** [comment]
yes, sure.
[profiling.zip](https://github.com/potentialventures/cocotb/files/3350166/profiling.zip)

# current master
current_master

# eric-wieser:avoid-runtime-coroutine-creation
avoid-runtime-cro-creation

**Comment by eric-wieser** [comment]
Sorry, what I meant by that is the same timing you did before

**Comment by andresdemski** [comment]
*  eric-wieser:avoid-runtime-coroutine-creation: SIM / REAL TIME : 13872.64 NS/S
* current master: SIM / REAL TIME : 13967.51 NS/S

No performance improvement.

**Comment by andresdemski** [comment]
*  eric-wieser:avoid-runtime-coroutine-creation: 21790.35 NS/S
* current master 2cd40a8: 22655.83 NS/S
*  eric-wieser:avoid-runtime-coroutine-creation after merging current master: 21995.51 NS/S

I found that the performance has a ~2% of standard deviation. So I need to do a better performance analisys tool.

**Comment by andresdemski** [comment]
Using https://github.com/andresdemski/cocotb-performance
```console
$ ./run_in_docker.sh python3 compare.py                                                                                                                                                                     
sha256:7a4d132d15ca92f95b7f75b637c2c95b277bf7a3609918815f276bbe9fb8c5e4

 url https://github.com/potentialventures/cocotb/
 commit: master
 path: /tmp/cocotb-master
 results: 
    simtime: 142610.10 ns
    realtime: 18.65s
    performance: 7648.49 ns/s


 url https://github.com/eric-wieser/cocotb/
 commit: avoid-runtime-coroutine-creation
 path: /tmp/cocotb-eric-wieser
 results: 
    simtime: 142610.10 ns
    realtime: 18.20s
    performance: 7833.63 ns/s


 url https://github.com/potentialventures/cocotb/
 commit: 2744c4fec607bf598d677484e6c0290183360f7a
 path: /tmp/cocotb-master
 results: 
    simtime: 142610.10 ns
    realtime: 15.53s
    performance: 9182.79 ns/s
```
performance improvement: ~2.4%

**Comment by imphil** [comment]
@eric-wieser so we can take this or do you plan further changes?

**Comment by eric-wieser** [comment]
I don't really see where 2.4 % comes from there, I read that delta as 17%.


Either way, the profiling svgs show this cuts the number of calls to coroutines by a factor of three, so I think this patch is a good idea.

I don't intend to make further changes.

**Comment by andresdemski** [comment]
2.4% comes from: 7833.63 ns/s / 7648.49 ns/s = 1.024 = 2.4%

9182.79 ns/s performace is using the commit i'm using at work (2744c4fec607bf598d677484e6c0290183360f7a)

**Comment by eric-wieser** [comment]
@andresdemski: Ah, I'd assumed that third commit was the merge of this branch with master.

**Pull request merged**